plasma_framework/contracts/src/exits/payment/controllers/PaymentChallengeIFENotCanonical.sol

```diff
@@ -0,0 +1,125 @@
+        IPaymentSpendingCondition condition = self.spendingConditionRegistry.spendingConditions(
+            args.competingTxInputOutputType, inFlightTx.txType
+        );
+        require(address(condition) != address(0), "Spending condition contract not found");
+
+        bool isSpentByInFlightTx = condition.verify(
+            ife.inputs[args.inFlightTxInputIndex].outputGuard,
+            uint256(0), // should not be used
+            inFlightTx.inputs[args.inFlightTxInputIndex],
+            args.competingTx,
+            args.competingTxInputIndex,
+            args.competingTxWitness
+        );
+        require(isSpentByInFlightTx, "Competing input spending condition is not met");
+
+        // Determine the position of the competing transaction
+        uint256 competitorPosition = ~uint256(0);
+        if (args.competingTxPos != 0) {
+            UtxoPosLib.UtxoPos memory utxoPos = UtxoPosLib.UtxoPos(args.competingTxPos);
+            (bytes32 root, ) = self.framework.blocks(utxoPos.blockNum());
+            competitorPosition = verifyAndDeterminePositionOfTransactionIncludedInBlock(
+                args.competingTx, utxoPos, root, args.competingTxInclusionProof
+            );
+        }
+
+        // Competitor must be first or must be older than the current oldest competitor.
+        require(
+            ife.oldestCompetitorPosition == 0 || ife.oldestCompetitorPosition > competitorPosition,
+            "Competing transaction is not older than already known competitor"
+        );
+
+        ife.oldestCompetitorPosition = competitorPosition;
+        ife.bondOwner = msg.sender;
+
+        // Set a flag so that only the inputs are exitable, unless a response is received.
+        ife.isCanonical = false;
+
+        emit InFlightExitChallenged(msg.sender, keccak256(args.inFlightTx), competitorPosition);
+    }
+
+    /**
+     * @dev Checks that in-flight exit is in phase that allows for piggybacks and canonicity challenges.
+     * @param ife in-flight exit to check.
+     */
+    function verifyFirstPhaseNotOver(
+        PaymentExitDataModel.InFlightExit storage ife,
+        uint256 minExitPeriod
+    )
+        private
+        view
+    {
+        uint256 phasePeriod = minExitPeriod / 2;
+        bool firstPhasePassed = ((block.timestamp - ife.exitStartTimestamp) / phasePeriod) >= 1;
+        require(!firstPhasePassed, "Canonicity challege phase for this exit has ended");
+    }
+
+    function verifyAndDeterminePositionOfTransactionIncludedInBlock(
+        bytes memory txbytes,
+        UtxoPosLib.UtxoPos memory utxoPos,
+        bytes32 root,
+        bytes memory inclusionProof
+    )
+        private
+        pure
+        returns(uint256)
+    {
+        bytes32 leaf = keccak256(txbytes);
+        require(
+            Merkle.checkMembership(leaf, utxoPos.txIndex(), root, inclusionProof),
+            "Transaction is not included in block of plasma chain"
+        );
+
+        return utxoPos.value;
+    }
+}
```

Comment:
Minor: GH usually argues on this for empty line in EOF....not a big deal though

Reply:
I will check what my editor has to say about ;)